### ISSUE TITLE
vcpkg upgrade process breaks installation if build process failed

### ISSUE THREAD

**Issue by FeyFre** [microsoft/vcpkg#20342]
**Describe the bug**
vcpkg incorrectly handles port upgrade failure. Current sequence is:
1 Uninstall old port
2 Download sources for new
3 Build new packages
4 Install build result
If steps 2 or 3 fail for any reason (network failure, insufficient disc space, user abort), we got broken environment without both old and new versions of port. Either some kind of rollback should happen (from backup) or irreversible destructive operations must be performed last. 

**Environment**
- OS: Windows
- Compiler: irrelevant

**To Reproduce**
1. ./vcpkg install xxxx
2. wait for some time for package xxxx to update
3. git pull to update ports info 
4. vcpkg update to see which ports are due to update
5. vcpkg upgrade --no-dry-run
6. During upgrade execution cause some issue so upgrade process fails (e.g. disconnect network during download of port source code)

*Alternative way to reproduce*
1. set environment variable VCPKG_FORCE_SYSTEM_BINARIES
2. delete cmake, git, other tools which are used during build process and usually downloaded automatically are deleted
3. Run vcpkg upgrade --no-dry-run

**Expected behavior**
I expected new port sources downloaded, built, prepared to be installed, and only after that old version to be uninstalled, and new version installed.

Alternative steps to reproduce yield the same issue: vcpkg will uninstall old version of packages, then attempts to build new, fails (missing tool, or because tools installed in system are unacceptable, in make case CMake was too new), and aborts leaving environment broken.

**Failure logs** 

**Additional context**
Pretty much everything is explained in *expected* section.

**Comment by autoantwort** [comment]
You can pass `--keep-going`

**Comment by PhoebeHui** [comment]
Closing in favor of https://github.com/microsoft/vcpkg/discussions/17922 https://github.com/microsoft/vcpkg/issues/4360.
